### ISSUE TITLE
[Bug] Activating borderless window mode in fullscreen crashes gw

### ISSUE THREAD

**Issue by AnonymousII** [HasKha/GWToolboxpp#123]
When you turn on tb borderless window mode while playing gw in Fullscreen, first of all nothing seems to happen, the mouse is still caught in gw and the close/minimize/window buttons in the top right corner wont disappear.
Now when you disable borderless window mode again, gw starts flashing ~5 times/sec on the screen obviously trying to switch between window mode and Fullscreen while taking focus all the time, so gw is effectively crashed (still running but you cant do anything in gw) and you have a hard time to regain control over your PC since you cant focus any windows or bring them to the front anymore. Only way to end it is close gw in the task bar with right click.

When you start gw again after closing it, screen resolution is reset to 800x600.

This bug is really annoying when you're almost done on a vanquish or doa run and you accidently turn on borderless window mode while you're in full screen (it works when you turn it on while in window mode).

If it takes too much time to make it work while in Fullscreen mode please at least check if we're in Fullscreen when turning on borderless window mode and display an error message in chat or simply do nothing - still better than crashing gw :)

**Comment by HasKha** [comment]
Turning borderless while in fullscreen doesn't and won't work. 

There is a check in place to see if the game is in fullscreen, but it's not too reliable. I'm not sure how to fix it. The proper fix would be to find fullscreen-enabled information in game memory, but we don't have that atm. 

Unfortunately my advice is just to not enable borderless while in fullscreen, sorry.

**Comment by AnonymousII** [comment]
Lol 3rd google hit on "c++ direct x check fullscreen of window" brought me to this stackoverflow post: [https://stackoverflow.com/questions/7009080/detecting-full-screen-mode-in-windows](url)

Heres a quick example how to reliably check if gw is running in Fullscreen (just tested, working really well):
bool isFullscreen(HWND window)
{
	RECT a, b;
	GetWindowRect(window, &a);
	GetWindowRect(GetDesktopWindow(), &b);
	return (a.left == b.left  &&
		a.top == b.top   &&
		a.right == b.right &&
		a.bottom == b.bottom);
}

HWND FindGW()
{
	return FindWindowA(NULL, "Guild Wars");
}

int main()
{
	HWND gw = FindGW();
	cout << "Handle: " << gw<<endl;

	Sleep(20000);

	cout << "IsFullscreen: " << isFullscreen(gw);
	_getch();
}

**Comment by HasKha** [comment]
The current check only relies on window style (https://github.com/HasKha/GWToolboxpp/blob/master/GWToolbox/GWToolbox/Modules/GameSettings.cpp#L230). Works for me, I wonder why it doesn't for you.

I don't think we can use the code you posted, it's only checking for gw to have the same resolution as the main desktop, which will be true under borderless window too.

**Comment by AnonymousII** [comment]
Well for the Fullscreen or borderless problem there is a really simple solution: Your tb is the only program that causes gw to go into borderless mode so you have full control of it...
-When tb starts up youre in Fullscreen or window mode (when tb is closed you can go back to window if in borderless)
-first time the user clicks borderless (you cant already be in borderless!) check resolution, if its desktop size youre in Fullscreen, otherwise its in window mode... When switching to borderless (cause you were in window mode) simply set a flag that tells you youre in borderless now...

But it's indeed strange that this doesnt work for me it actually looks pretty decent... I'll debug this code part within next days cause im also interested in why it doesnt work for me and report back ;)

**Comment by HasKha** [comment]
keep in mind that you can close and re-open toolbox with gw in borderless. This is intentional so people can quickly use tb to set borderless and don't have to keep it open. 

The flag that you are talking about, rather than being stored in toolbox memory, is the WS_POPUP style that toolbox sets to the gw window.

We could check for it (e.g. if resolution = fullscreen and no popup style -> fullscreen), however, since the style check fails for you, I wonder if this would fail as well.

**Comment by AnonymousII** [comment]
ok so we Need  to distinguish between window rect and client rect:
Window rect: [https://msdn.microsoft.com/de-de/library/windows/desktop/ms633519(v=vs.85).aspx](url)
Client rect: [https://msdn.microsoft.com/de-de/library/windows/desktop/ms633503(v=vs.85).aspx](url)

So we can clearly distinguish between window, Fullscreen and borderless:

- if GetWindowRect(GW)==GetWindowRect(Desktop) its either Fullscreen or borderless
- - if GetClientRect(gw)!=GetWindowRect(gw) theres still a window border just moved out of sight (as you do for borderless) --> were in borderless
- - else were in Fullscreen
- else were in window mode

Still untested but i guess it should work like that -- ill test it in the afternoon ;)

**Comment by AnonymousII** [comment]
Ok just tested it and found a nice solution that has 100% hit rate (even in different screen resolutions) and can distinguish between window, Fullscreen and borderless :)
We need a combination of your check for WS_MAXIMIZEBOX-Style and screen resolution:
- When gw resolution = desktop resolution were in Fullscreen or borderless
  - Full Screen removes WS_MAXIMIZEBOX, so if WS_MAXIMIZEBOX is no set its Fullscreen
  - else its borderless
- else its window



Here's the code for the window/borderless/Fullscreen check:

```
RECT windowRect, desktopRect;
GetWindowRect(gw, &windowRect);
GetWindowRect(GetDesktopWindow(), &desktopRect);

```


		//fullscreen or borderless
		if (RectEquals(windowRect, desktopRect))
		{
			DWORD current_style = GetWindowLong(gw, GWL_STYLE);

			if ((current_style & WS_MAXIMIZEBOX) != 0)
			{
				cout << "Borderless\n";
			}
			else
			{
				cout << "Fullscreen\n";
			}
		}
		else
		{
			cout << "Window\n";
		}

While RectEquals just compares the bounds:
`inline bool RectEquals(RECT a, RECT b) {return (a.left == b.left  && a.top == b.top   && a.right == b.right && a.bottom == b.bottom);}`

And FindGw gets gw window handle:
`inline HWND FindGW() { return FindWindowA(NULL, "Guild Wars");}`

works like a charm, easy & safe way to distinguish between full screen and borderless ;)

**Comment by HasKha** [comment]
Thanks. You essentially replaced the WS_POPUP check with the resolution-based one. Looks good, although I still wonder:
- Why does WS_POPUP check not work?
- This setup will fail in multi-monitor setups where the user resized the gw window after setting borderless (for example, for having gw spanning multiple monitors). 
So overall we should probably have both checks (if RectEquals(..) || WS_POPUP style). 

Also minor but it's better to use `GW::MemoryMgr::GetGWWindowHandle()` to get the window handle :)

**Comment by AnonymousII** [comment]
> •Why does WS_POPUP check not work?

There are a few interesting observations I made while experimenting with window styles in GW:
- Fullscreen always has excactly WS_CLIPSIBLINGS | WS_VISIBLE for me no matter what combination of window / Fullscreen / borderless I did before and no matter if Toolbox is running or not --> this explains why WS_POPUP-Check fails: for Borderless you manually set WS_POPUP, but as soon as you change to Fullscreen by any combination it gets removed. (Win 10 Pro 64 bit, Version 1703 Build 15063.413)
-Whenever you do the combination "window -> Fullscreen -> window" WS_POPUP gets added to the window style (usually not in window style) - this means you definitely cant rely on WS_POPUP not being set in window mode! (when you use borderless -> window afterwards WS_POPUP is removed again since you explicitly set the style in tb code)

> This setup will fail in multi-monitor setups where the user resized the gw window after setting borderless (for example, for having gw spanning multiple monitors).
>  So overall we should probably have both checks (if RectEquals(..) || WS_POPUP style).

At least in Win 10 I can't find a way a regular user could resize gw so that it spawns multiple monitors - when in window mode and i pull the border over to the other monitor it bounces back and in borderless mode i can't resize the window at all because toolbox removes the WS_THICKFRAME-Style which gives you the resize-cursor. 

Of course with some programming skills I can resize the window with MoveWindow() while in borderless mode but i dont see a way how the average toolbox user should accomplish this...

Nevertheless if you have a way to make gw spawn 2 monitors in borderless mode the code above only needs a small extension, really easy to catch: If the window rect width or height is bigger than screen resoultion youre in borderless with multi-monitor gw (theres absolutely no way in win 10 to get gw to spawn multiple monitors in window mode, even with some coding skills like MoveWindow(), SetWindowLong(), just tested...)

So if you wanna support this scenario as well all we need is a little extension ;)

```
//fullscreen or borderless
		if (RectEquals(windowRect, desktopRect))
		{
			DWORD current_style = GetWindowLong(gw, GWL_STYLE);

			if ((current_style & WS_MAXIMIZEBOX) != 0)
			{
				cout << "Borderless\n";
			}
			else
			{
				cout << "Fullscreen\n";
			}
		}
		else if (RectMultiscreen(desktopRect, windowRect))
		{
			cout << "Borderless Multiscreen\n";
		}
		else
		{
			cout << "Window\n";
		}
```

No idea how the default user could get borderless mode to spawn multiple monitors without coding skills but its really easy to integrate it in the resolution-based check - now this scenario works as well (just tested it with manually making borderless mode multiscreen using MoveWindow() ;) )

**Comment by HasKha** [comment]
Thanks for the thorough testing. The WS_POPUP behavior is surprising. 

Yes, you need a simple program to resize gw while in borderless, but that can be achieved with a 1-line AutoIt script, which I suggested to a few people and it's easy to run and customize. I'm just making sure to not break anyone's workflow. Your last suggestion looks great and should handle this. 

Feel free to make a PR if you want, otherwise I'll add it myself soon-ish. Again, we need to change the check at https://github.com/HasKha/GWToolboxpp/blob/master/GWToolbox/GWToolbox/Modules/GameSettings.cpp#L230 with the RectEquals-based check.

**Comment by AnonymousII** [comment]
Ok found the last part of the puzzle as well :D

borderless has these styles:
WS_CLIPSIBLINGS | WS_GROUP | WS_MAXIMIZEBOX | WS_MINIMIZEBOX | WS_POPUP | WS_TABSTOP | WS_VISIBLE

window those:
WS_BORDER | WS_CAPTION | WS_CLIPSIBLINGS | WS_DLGFRAME | WS_GROUP | WS_MAXIMIZEBOX | WS_MINIMIZEBOX | WS_SIZEBOX | WS_TABSTOP | WS_THICKFRAME | WS_VISIBLE | (WS_POPUP)

so window has these styles borderless doesnt have:
WS_BORDER | WS_CAPTION | WS_DLGFRAME | WS_GROUP | WS_SIZEBOX | WS_THICKFRAME 

I'd use WS_BORDER and WS_THICKFRAME since they will for sure not be used in borderless ;)

So the full code that can distinguish between all possible scenarios is this:

```
//fullscreen or borderless
		if (RectEquals(windowRect, desktopRect))
		{
			
			if ((current_style & WS_MAXIMIZEBOX) != 0)
			{
				cout << "Borderless\n";
			}
			else
			{
				cout << "Fullscreen\n";
			}
		}
		else if (RectMultiscreen(desktopRect, windowRect))
		{
			cout << "Borderless Multiscreen\n";
		}
		else if ((current_style & WS_BORDER) == 0 && (current_style & WS_THICKFRAME) == 0)
		{
			cout << "Borderless with window size smaller than Desktop\n";
		}
		else
		{
			cout << "Window\n";
		}
```

Now you really got all possible scenarios covered even if the user has some coding skills and resizes the gw window using WIn32-API ;)  (Although I still can't think of a way to regularly resize borderless mode...)

**Comment by AnonymousII** [comment]
ok made a pull request with the updated code - now Fullscreen check should always work ;)

While debugging i noticed some strange behaviour: When starting Toolbox while gw is in windod mode, GameSettings::ApplyBorderless() gets called twice on tb startup with parameter false. Since this would trigger the info message I temporarily commented it out so users wont get confused by strange tb internal behaviour. Nevertheless we should check where these calls come from and possibly fix it - they dont really seem to make sense, I'll debug this tomorrow ;)

**Comment by HasKha** [comment]
The reason for the double call is in GWToolbox.cpp, lines 283 and 287, where `LoadModuleSettings `is called twice. Loading settings twice should never be a problem, but I'm fixing it, to avoid confusion like this in the future and improve performance (although imperceptible).

**Comment by AnonymousII** [comment]
Alright, thanks for merging :)

For the settings-load: of course loading twice should work (and in fact it does work, we only get a warning that we are already in window mode). I was just surprised that we don't check whether we're already in window mode and thus try to set window mode while we're already in window mode...
Actually we should check current game status and only apply settings if they are not equal to current game status - So we just put the window/borderless/Fullscreen check in a function and call that on settings load ;)

Maybe I'll implement it in a few days and make another PR - or we just silently ignore these warnings, then we should remove them all...

Btw just saw that i forgot to remove a debug message in https://github.com/HasKha/GWToolboxpp/blob/master/GWToolbox/GWToolbox/Modules/GameSettings.cpp#L239  - just delete it not to confuse tb users :)

**Comment by HasKha** [comment]
well I changed it, it doesn't double load anymore. 

The current behavior is what we want. ApplyBorderless *is* that function, it changes the state of the game to its parameter if possible, or prints a warning if not. It cannot be a simple toggle, since the user is allowed to `/borderless on` when already enabled. 

I changed the debug message from a Log::Info to a Log::Log, so it doesn't appear in-game but is sent to console or log file, which is ok. I might remove it later after some testing.

**Comment by AnonymousII** [comment]
> ApplyBorderless _is_ that function, it changes the state of the game to its parameter if possible, **or prints a warning if not**

Well I just thought the warning would confuse the user when he starts tb in window mode and gets the warning "Already in window mode" without doing anything, so I wouldn't print that warning if the mode change is a result of loading settings - but whatever if printing the warning anyway is the intended behaviour we don't need to change anything, finally its your decision...

**Comment by HasKha** [comment]
closing, implemented in #124 (thanks again). I *think* the current notification behavior is ok, it always prints a message except when turning off when it's already off. I'm not a fan of having the exception, but it's no big deal.
